### ISSUE TITLE
Reponse body io.Reader

### ISSUE THREAD

**Issue by qJkee** [valyala/fasthttp#411]
Hey guys. How to set custom response body writer?
I've tried to use `SetBodyStream`, but i don't see any body readers from response which i need to pass in `io.Reader`.
P.S I'm trying to implement throttler.

**Comment by qJkee** [comment]
Problem is, that I can't get/reach `Responce.bodyStream`

**Comment by erikdubbelboer** [comment]
Here is a simple example that throttles the response to 2 bytes per second:
```go
package main

import (
	"bufio"
	"io"
	"net/http"
	"time"

	"github.com/valyala/fasthttp"
)

func rateLimited(ctx *fasthttp.RequestCtx) {
	body := []byte("this is a test")

	ctx.SetBodyStreamWriter(func(w *bufio.Writer) {
		i := 0
		for i < len(body) {
			end := i + 2
			if end >= len(body) {
				end = len(body) - 1
			}

			w.Write(body[i:end])
			w.Flush()

			i += 2

			time.Sleep(time.Second)
		}
	})
}

func main() {
	go func() {
		if err := fasthttp.ListenAndServe(":8080", rateLimited); err != nil {
			panic(err)
		}
	}()

	res, err := http.Get("http://localhost:8080")
	if err != nil {
		panic(err)
	}

	for {
		var b [2]byte

		if _, err := res.Body.Read(b[:]); err == io.EOF {
			break
		} else if err != nil {
			panic(err)
		}

		println(string(b[:]))
	}
}
```

**Comment by qJkee** [comment]
It's correct.
But i need to read the body from `fasthttp.Response`(i've done request to another website), not from static one

**Comment by qJkee** [comment]
@erikdubbelboer 
?

**Comment by erikdubbelboer** [comment]
@qJkee Just use [`Response.Body()`](https://godoc.org/github.com/valyala/fasthttp#Response.Body)

**Comment by qJkee** [comment]
@erikdubbelboer 
It will return me `[]byte`, not `io.Reader`

**Comment by erikdubbelboer** [comment]
@qJkee look at my example! it's not using any `io.Reader`.

**Comment by dgrr** [comment]
I think @qJkee wants to make client request and read the response from io.Reader. And this is not possible using fasthttp.

**Comment by erikdubbelboer** [comment]
@qJkee if that is what you want I'm afraid @dgrr is correct that it's not possible. Fasthttp always reads the whole response body before returning.

**Comment by erikdubbelboer** [comment]
@qJkee because this requires huge internal changes. The way things work internally in fasthttp wasn't designed for this at all.

To demonstrate this I actually wrote a version that would add support for this with backwards compatibility and keeping it possible to work with 0 heap allocations.

You can find my changes here: https://github.com/erikdubbelboer/fasthttp/commit/69515271036c791b25543da6a4360fadb6b61173

It's quite ugly and complicated code. It introduces 2 more `sync.Pool` uses with each request. And it increases the coupling between the different components which I don't like that much either.

In the current state the API could also use some improvements. Right now you have to set `Response.ReturnBodyReader` to `true` in which case after the request `Response.BodyReader` contains an `io.ReadCloser` which allows you to read the body.

Just like `net/http` it requires you to read the whole body and call `Response.BodyReader.Close()` for the connection to be reused for future request. On top of that right now it also requires calling `ReleaseBodyReader(Response.BodyReader)` but I think this can be merged into the `Response.BodyReader.Close` call.

A [full example can be found in the benchmarks](https://github.com/erikdubbelboer/fasthttp/blob/clientbodyreader/erik_test.go#L70-L95) that I added to make sure I introduces no new heap allocations.

When I have time in the future I might improve this and at some point merge it into the master branch.

**Comment by erikdubbelboer** [comment]
I think I know what is happening. You are trying to read a buffer that is bigger than the current chunk and it hangs trying to read those extra bytes (which only arrive at the next chunk).

I have changed my implementation a bit so `Read()` now returns as soon as a chunk is finished. This way the following example works fine for me:
```go
package main

import (
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/erikdubbelboer/fasthttp"
)

func readBody(resp *fasthttp.Response) {
	// 40 bytes is bigger than the chunk of 9 we're about to receive.
	b := make([]byte, 40)

	for {
		if n, err := resp.BodyReader.Read(b); err != nil {
			break
		} else {
			log.Printf("body: %q", b[:n])
		}
	}
}

func doRequest() {
	client := &fasthttp.Client{
		DialDualStack:                 true,
		DisableHeaderNamesNormalizing: true,
	}

	req := fasthttp.AcquireRequest()
	req.Header.SetMethod("GET")
	req.SetRequestURI("http://localhost:8080/test")

	resp := fasthttp.AcquireResponse()
	resp.ReturnBodyReader = true

	err := client.Do(req, resp)
	if err != nil {
		panic(err)
	}

	go readBody(resp)

	log.Println("after go readBody")
}

func main() {
	http.HandleFunc("/test", func(w http.ResponseWriter, r *http.Request) {
		flusher, ok := w.(http.Flusher)
		if !ok {
			panic("expected http.ResponseWriter to be an http.Flusher")
		}
		for i := 1; ; i++ {
			fmt.Fprintf(w, "Chunk #%d\n", i)
			flusher.Flush() // Trigger "chunked" encoding and send a chunk...
			time.Sleep(500 * time.Millisecond)
		}
	})

	go func() {
		log.Print("Listening on localhost:8080")
		log.Fatal(http.ListenAndServe(":8080", nil))
	}()

	// Give the server 100ms to start.
	time.Sleep(time.Millisecond * 100)

	doRequest()

	// Exit the program after 10 seconds.
	time.Sleep(time.Second * 10)
}
```

**Comment by carlonelong** [comment]
OK. It works. And here comes up another question. How could I set the proper buffer size... ？ I just want to use some code like this:
```go
reader := bufio.NewReader(data) // I have to use bufio.NewReaderSize(data, 100) instead
for {
            fmt.Printf("decoding %#v\n", reader)
            lengthStr, err := reader.ReadString('\n')
            fmt.Printf("find a ln\n")
            if err != nil {
                    return err
            }
            //  do something.
}
```

Background is that this "bufio.NewReader" use http.Response before, and I want to change it to fasthttp.Response.
Thanks!

**Comment by erikdubbelboer** [comment]
You don't really need to set an initial buffer size. Let the reader figure that out itself. Your code works fine. Try replacing the function in my example with this:
```go
func readBody(resp *fasthttp.Response) {
  reader := bufio.NewReader(resp.BodyReader)

  for {
    s, err := reader.ReadString('\n')
    if err != nil {
      break
    } else {
      log.Printf("body: %q", s)
    }
  }
}
```

**Comment by erikdubbelboer** [comment]
The problem you had was fixed when I introduced the commit that makes `Read` return when a chunk is finished.

**Comment by carlonelong** [comment]
Sorry. Missed this. 
It now works fine. Thanks!

**Comment by Arnold1** [comment]
> Fasthttp always reads the whole response body before returning

is that still true, or is there a way to use read the response body with io.Reader?

**Comment by erikdubbelboer** [comment]
@Arnold1 I'm afraid that this is still true, the code I wrote above to add support for this isn't fit to be merged and I'm not sure if I could ever get it good enough. Reading the whole body is just how fasthttp was designed I'm afraid.

**Comment by Arnold1** [comment]
whats the best way to get high thoughput + low latency with fasthttp?
i have a POST handler and have a function which takes 5ms for calculation than i send the response back...if the calculation takes longer than 20ms i need to send back a http.StatusRequestTimeout

**Comment by erikdubbelboer** [comment]
To set server options use:
```go
	s := &fasthttp.Server{
		Handler:               router.Handler,
		ReadTimeout:           time.Hour,
		WriteTimeout:          time.Hour,
		ReadBufferSize:        4096 * 6,
		WriteBufferSize:       4096 * 6,
		NoDefaultServerHeader: true,
	}
	log.Fatal(s.ListenAndServe(":8080"))
```
You probably don't have to change the concurrency as the default is `262144` concurrent connections which should always be enough.

If you really want to have the best performance I would suggest not using fasthttprouter but doing your own simple routing. Unless of course your routing is complex in which case fasthttprouter is the better option.

For JSON I would suggest using https://github.com/mailru/easyjson for best performance.

Your timeout code doesn't make sense this way, it needs to be something like this:
```go
	done := make(chan struct{})

	// Preallocate the correct slice capacity already for optimal performance.
	allResp := make([]*response, 0, len(modelData))
	
	// Do the processing in a goroutine so we can abandon it if it takes too long.
	go func() {
		defer close(done)

		// In theory you could also do each calcResponse in parallel goroutines
		// to speed it up even more if its slow.
		for _, model := range modelData {
			// do calculation ... 5-7ms, but can be called multiple times, bc i range over modelData...
			response := calcResponse(jsonData)
			allResp = append(allResp, response)
		}
	}()

	// Either wait until the processing is done, or the timeout.
	select {
	case done:
	case <-time.After(30 * time.Millisecond):
		ctx.SetStatusCode(fasthttp.StatusRequestTimeout)
		respBody, _ := json.Marshal(errMessage)
		ctx.Write(respBody)
		return // If we timeout we should return here and not send back the normal response.
	}
	
	ctx.Write(allResponseJSON)
```

**Comment by Arnold1** [comment]
hi, why do you have 2 ctx.Write? ctx.Write(respBody) and ctx.Write(allResponseJSON)?

**Comment by Arnold1** [comment]
hi can you do the following without fasthttprouter?
```
router := fasthttprouter.New()
router.POST("/getResponse", LogWrapper(HTTPErrorWrapper(calcResp)))
```

**Comment by erikdubbelboer** [comment]
Just check the paths in your one handler:
```go
var (
	strSlashGetResp = []byte("/getResp")
)

func handler(ctx *fasthttp.RequestCtx) error {
	if bytes.Equal(ctx.Path(), strSlashGetResp) {
		// ...
	}
}
```

**Comment by Arnold1** [comment]
@erikdubbelboer how can i run the ```for _, model := range modelData {``` in parallel?

**Comment by erikdubbelboer** [comment]
Something like:
```go
	done := make(chan struct{}, len(modelData))

	// Preallocate the correct slice capacity already for optimal performance.
	allResp := make([]*response, 0, len(modelData))
	var allRespMu sync.Mutex

	// In theory you could also do each calcResponse in parallel goroutines
	// to speed it up even more if its slow.
	for _, model := range modelData {
		go func(model) {
			defer func() {
				done<-struct{}{}
			}()

			// do calculation ... 5-7ms, but can be called multiple times, bc i range over modelData...
			response := calcResponse(jsonData)

			allRespMu.Lock()
			allResp = append(allResp, response)
			allRespMu.Unlock()
		}(model)
	}

	// Either wait until the processing is done, or the timeout.
	for _ := range modelData {
		select {
		case <-done:
		case <-time.After(30 * time.Millisecond):
			ctx.SetStatusCode(fasthttp.StatusRequestTimeout)
			respBody, _ := json.Marshal(errMessage)
			ctx.Write(respBody)
			return // If we timeout we should return here and not send back the normal response.
		}
	}
	
	allRespMu.Lock()
	ctx.Write(allResp)
```

Or you could even use the channel to pass down the results instead of using a Mutex. That would be better but I don't have the time to write that now.

**Comment by Arnold1** [comment]
is there any issue with i do the following?
```
cancelFunc, cancel := context.WithTimeout(ctx, time.Duration(30)*time.Millisecond) // ctx from fasthttp ... ctx *fasthttp.RequestCtx
	defer cancel()
	resultsCh := make(chan *format.ResponseEntity, len(entityModel))
	results := make([]*format.ResponseEntity, len(entityModel))

	for _, model := range modelData {
		go func(model *ModelData) {
			// do calculation
			response := calcResponse(jsonData)

			resultsCh <- &ResponseEntity{
				Results:           response,
				MetaData:       model.ModelID,
				// and other things...
			}

		}(model)
	}

	i := 0
	for {
		select {
		case <-cancelFunc.Done():
			status = fasthttp.StatusRequestTimeout // 408

                        // i might want to return partial data...
			/*
                       ctx.Write(Results{Results: results[0:i]})
                        */
			return nil
		case r := <-resultsCh:
			results[i] = r
			i++
			if i == len(modelData) {
				ctx.Write(Results{Results: results})
				return nil
			}
		}
	}
```

**Comment by erikdubbelboer** [comment]
Arnold1 Yes that's perfect.

**Comment by Arnold1** [comment]
for some reason the the code works slower...hm...

**Comment by Arnold1** [comment]
@erikdubbelboer 
what is the diff between:

1.)
```
func calcResponse(ctx *fasthttp.RequestCtx) {
   cancelFunc, cancel := context.WithTimeout(ctx, time.Duration(30)*time.Millisecond)
   //...
}
```
2.
```
func calcResponse(ctx *fasthttp.RequestCtx) {
   cancelFunc, cancel := context.WithTimeout(context.Background(), time.Duration(30)*time.Millisecond)
   //...
}
```

**Comment by erikdubbelboer** [comment]
In case of 1 your context will also be cancelled if you shut down the server using [server.Shutdown()](https://godoc.org/github.com/valyala/fasthttp#Server.Shutdown).

**Comment by geraldstanje** [comment]
@erikdubbelboer is there any obvious reason why mutex version works faster than the context version i posted before?

**Comment by erikdubbelboer** [comment]
In theory the channel version should be slightly faster as channels can be slightly faster for the scheduler. But in practice they are probably the same speed.

**Comment by Arnold1** [comment]
@erikdubbelboer 

if modelData is size 3, i need to wait for 3 done to be received. but here it seems to exit the foor loop after the first done?
```
// Either wait until the processing is done, or the timeout.
	select {
	case done:
	case <-time.After(30 * time.Millisecond):
		ctx.SetStatusCode(fasthttp.StatusRequestTimeout)
		respBody, _ := json.Marshal(errMessage)
		ctx.Write(respBody)
		return // If we timeout we should return here and not send back the normal response.
	}
```

**Comment by erikdubbelboer** [comment]
Use: https://golang.org/pkg/testing/#hdr-Benchmarks

In that case just assign the `time.After(30 * time.Millisecond)` to a variable outside the loop so it is only started once.

**Comment by geraldstanje** [comment]
@erikdubbelboer cool, do you have a sample where you benchmark fasthttp code? should i send a request to /Response endpoint and benchmark that?

**Comment by Arnold1** [comment]
@erikdubbelboer cool, do you have a sample where you benchmark fasthttp code? should i send a request to /Response endpoint and benchmark that?

**Comment by erikdubbelboer** [comment]
There is bunch of test code in here for example: https://github.com/valyala/fasthttp/blob/e5f51c11919d4f66400334047b897ef0a94c6f3c/server_timing_test.go

**Comment by Fenny** [comment]
I think his question has been answered with examples and links to specific resources, can we close this? @erikdubbelboer

**Comment by karelbilek** [comment]
When I set `resp.StreamBody = true`, the `resp.BodyStream()` is an io.reader that has the response body. 

The API of fasthttp is very confusing here, but I think it's a "response body io reader"? Is this what this MR tried to do?
